src/thread/n3ds/SDL_syssem.c

```diff
@@ -72,18 +72,17 @@ int SDL_SemWaitTimeoutNS(SDL_sem *sem, Sint64 timeoutNS)
         return 0;
     }
 
-    if (timeoutNS == 0) {
+    {
         int result = LightSemaphore_TryAcquire(&sem->semaphore, 1);
-        /* TryWait will block unless we wait about 50 microseconds */
-        SDL_DelayNS(SDL_US_TO_NS(50));
-        return result == 0 ? 0 : SDL_MUTEX_TIMEDOUT;
-    }
+        if (result != 0 && timeoutNS != 0) {
+            result = WaitOnSemaphoreFor(sem, timeoutNS);
+        }
 
-    if (LightSemaphore_TryAcquire(&sem->semaphore, 1) != 0) {
-        return WaitOnSemaphoreFor(sem, timeoutNS);
-    }
+        /* Avoid starvation since this is a cooperative threading model */
```

Comment:
I'm not sure we want an unconditional sleep here. It would introduce unnecessary delay if the semaphore is always signaled. If the timeout is 0, the application explicitly does not want to delay.

Reply:
The real problem is that this test is designed to test behavior with preemptive multitasking. 0% contention here isn't really a valid result.

Hmm, okay, can you very clearly document what's happening and why you're doing what you're doing?

Reply:
I'm surprised that a timeout != 0 ends up starving threads. Wouldn't that translate into a cooperative yield for that amount of time?

Reply:
Apologies for the delay, here is a situation which I think illustrates the pros and cons of this approach.

First a bit of background on the 3DS's threading model. It's cooperative, meaning the thread must manually yield for another one to run. Userland processes can only run on a single core. And lastly, it seems that at equal priority, scheduling is done in a round-robin fashion.

For the example, we have one semaphore that is roughly equivalent to a thread-safe queue of data. One or more thread that append data to the queue, once the queue is full and/or they have no data left to add, they sleep. On the other hand, we have N threads consuming data from the queue. The consumer threads have three ways of waiting for data to be available:

- A blocking wait:
  ```c
  while(alive) {
    void *data = Acquire(queue);
    DoWork(data);
  }
  ```
- A busy wait:
  ```c
  while(alive) {
    void *data = TryAcquire(queue);
    if (data != NULL) {
      DoWork(data);
    }
  }
  ```
- A wait with timeout:
  ```c
  while(alive) {
    void *data = TryAcquireTimeout(queue, timeout);
    if (data != NULL) {
      DoWork(data);
    }
  }
  ```

There is one current issue with the busy wait. Whether there is data or not, the thread will never yield, starving all other threads from ever running again. To address this, the first idea is to yield execution if the queue was empty. This avoids locking up the system by running this thread indefinitely. But with that fix alone, we have a new problem, the first thread after a producer empties the queue and starves the other consumer threads.

There is a similar issue with the timeout wait. It will only sleep if there are no data, and the first consumer thread will starve the others.

To address this issue, we introduce an unconditional yield at the end of `TryAcquire` and `TryAcquireTimeout`. This spreads the data equally over each thread.

This has a few caveats:

- If there is only one consumer thread, this brings an important runtime cost.
- Potentially unexpected behaviour, when you try to acquire data with no timeout, you don't necessarily expect your thread to sleep.

As for the tl;dr:

- By introducing a yield in `TryWait` if we fail, we avoid monopolising the CPU time indefinitely.
- By introducing an unconditional yield in `TryWait` and `TryWaitTimeout`, we get the benefit from above; plus we avoid starving the other threads at the cost of higher runtime overhead.

This latter change allows for a more predictable behaviour. Preemptive systems would (likely) spread the work more or less equally on their thread, so someone using SDL would expect this to be true on supported platforms. This change emulates this behaviour, but I can totally understand the point where the runtime cost is too high, that this may be too unlikely of a situation, or if this unnecessary.

EDIT: I've noticed when re-reading that doing a blocking wait would also starve other consumer threads.

Reply:
At this point, if you document it well, I'm fine with you committing the solution you're comfortable with.